### PR TITLE
Add bindings for nsCSSShadowArray, use for text-shadow and box-shadow

### PR DESCRIPTION
**Pull request by Manishearth**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13086)

### CODE REVIEW

**Comment by Manishearth** [comment]
Gecko bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1298655

File: ports/geckolib/gecko_bindings/structs_release.rs [hunk 1]
```diff
@@ -5895,14 +5895,13 @@ fn bindgen_test_layout_nsCSSShadowItem() {
...
-    pub mRefCnt: nsAutoRefCnt,
-    pub _mOwningThread: nsAutoOwningThread,
+    pub mRefCnt: ThreadSafeAutoRefCnt,
     pub mLength: u32,
     pub mArray: [nsCSSShadowItem; 1usize],
 }
 #[test]
 fn bindgen_test_layout_nsCSSShadowArray() {
-    assert_eq!(::std::mem::size_of::<nsCSSShadowArray>() , 48usize);
+    assert_eq!(::std::mem::size_of::<nsCSSShadowArray>() , 40usize);
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/structs_release.rs]: Hmm... why do we have owning threads in the release bindings?

File: ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ use std::ops::{Deref, DerefMut};
+ use std::{ptr, slice};
+
+ impl RefPtr<nsCSSShadowArray> {
+    pub fn replace_with_new(&mut self, len: u32) {
+        unsafe {
+            if !self.mRawPtr.is_null() {
+                Gecko_ReleaseCSSShadowArrayArbitraryThread(self.mRawPtr);
+            }
+            if len == 0 {
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: nit: This probably looks cleaner like:

```rust
self.mRawPtr = if len == 0 {
    ptr::null_mut()
} else {
    Gecko_Construct_...
};
```
```

File: ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                Gecko_AddRefCSSShadowArrayArbitraryThread(other.mRawPtr);
+            }
+
+            self.mRawPtr = other.mRawPtr;
+        }
+    }
+ }
+
+ impl Deref for RefPtr<nsCSSShadowArray> {
+    type Target = [nsCSSShadowItem];
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: nit: Indentation is off here.

Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: And in a few more places indeed.

**Comment by Manishearth** [comment]
screen shot 2016-08-28 at 11 49 58 am

File: ports/geckolib/gecko_bindings/structs_release.rs (hunk 1 above)
Comment (emilio) [inline ports/geckolib/gecko_bindings/structs_release.rs]: NVM, they're not exclusive.

File: ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+ use bindings::Gecko_Construct_CSSShadowArray;
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: Somehow all the lines here still have preceding whitespace.

File: ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ }
+
+impl Deref for RefPtr<nsCSSShadowArray> {
+    type Target = [nsCSSShadowItem];
+    fn deref(&self) -> &[nsCSSShadowItem] {
+        if self.mRawPtr.is_null() {
+            &[]
+        } else {
+            unsafe {
+                slice::from_raw_parts(&(*self.mRawPtr).mArray[0] as *const _,
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: `mArray.as_ptr()`?

File: ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+ }
+
+impl DerefMut for RefPtr<nsCSSShadowArray> {
+    fn deref_mut(&mut self) -> &mut [nsCSSShadowItem] {
+        if self.mRawPtr.is_null() {
+            &mut []
+        } else {
+            unsafe {
+                slice::from_raw_parts_mut(&mut (*self.mRawPtr).mArray[0] as *mut _,
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: ditto

**Comment by emilio** [comment]
r=me when this is ready.

File: ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                Gecko_ReleaseCSSShadowArrayArbitraryThread(self.mRawPtr);
+            }
+            if !other.mRawPtr.is_null() {
+                Gecko_AddRefCSSShadowArrayArbitraryThread(other.mRawPtr);
+            }
+
+            self.mRawPtr = other.mRawPtr;
+        }
+    }
+ }
```
Comment (emilio) [inline ports/geckolib/gecko_bindings/sugar/ns_css_shadow_array.rs]: nit: Still whitespace :-)

**Comment by Manishearth** [comment]
@bors-servo r=emilio

https://bugzilla.mozilla.org/show_bug.cgi?id=1298655 landed

**Comment by Manishearth** [comment]
@bors-servo retry
  - #11574 
  - #13037

**Pull request merged**